hail/src/main/scala/is/hail/asm4s/Code.scala

```diff
@@ -923,15 +933,21 @@ class LocalRefInt(val v: LocalRef[Int]) extends AnyRef {
 }
 
 class FieldRef[T, S](f: reflect.Field)(implicit tct: ClassTag[T], sti: TypeInfo[S]) {
```

Comment:
I'm confused by the relationship between `ClassFieldRef` and `FieldRef`. It appears that `FieldRef` is only ever used for static fields. If so, maybe we should specialize this to `StaticFieldRef` and simplify it, in which case I think this could be a `Settable` since it doesn't need the object ref. If that's out of scope for this PR, that's fine.

Reply:
That all makes sense. Thanks for explaining.

> I was playing with newFieldThisRef but ugh.

Do you agree that the difference between locals and fields on the `FunctionBuilder` object should eventually not be visible in the `Code` interface? If `FunctionBuilder`, `MethodBuilder`, etc. (or what they evolve into) are only used in the `Code` -> lir emitter, that might make naming less of an issue.

Reply:
> Do you agree that the difference between locals and fields on the FunctionBuilder object should eventually not be visible in the Code interface?

I'm not totally sure what you mean, but I think no.  I think treating fields like locals is completely an artifact of splitting and will go away when splitting is handled by lir.  I want FunctionBuilder to be a wrapper for a ClassBuilder[AsmFunctionN] plus an apply method builder, and that's it.  Once wrapping is gone, the code generator should use locals unless it explicitly wants a field (for whatever reason).  I would want field access to look like field access in a normal language, so `mb(f)` or `obj(f)` or `mb.self(f)` where `f` is a string or has been returned from `ClassBuilder.newField`.  The latter two expressions being the same.  A method access should be Value[T] if the object is, and Code[T] otherwise ... although that means `obj.a.b.c.d.e...zzz` is a Value[T], which doesn't seem great.  This is one place where SimpleValue[T] would be useful.

To make the wrapping interface clearer, I imagine FunctionBuilder, MethodBuilder, and ClassBuilder will have corresponding traits FunctionBuilderLike, etc.  If you have a pointer to a class builder (like method and function builder), then you are ClassBuilderLike.  So MethodBuilder will get you the class interface, and FunctionBuilder will get the method interface (for the apply method).